src/workerd/api/sockets.c++

```diff
@@ -137,13 +152,27 @@ void PipelinedAsyncIoStream::getpeername(struct sockaddr* addr, uint* length) {
 
 kj::Promise<size_t> PipelinedAsyncIoStream::read(void* buffer, size_t minBytes, size_t maxBytes) {
   return thenOrRunNow<size_t>([=](kj::Own<kj::AsyncIoStream>* stream) {
-    return (*stream)->read(buffer, minBytes, maxBytes);
+    auto& context = IoContext::current();
+    return context.awaitJs(context.awaitIo(
+        (*stream)->read(buffer, minBytes, maxBytes), [this](size_t size) {
+      if (size == 0) {
```

Comment:
A zero-size read indicates EOF but does not indicate that the socket is "closed"; the write direction may still be active, and many protocols rely on the ability to continue writing after receiving an EOF. Does the spec really say that we should treat EOF as a close event?

Reply:
The semantics that Node.js follows here are essentially this:

* Every socket has an `allowHalfOpen` flag that is `false` by default.

* When the socket *receives* EOS, if `allowHalfOpen` is `false`, the writable side of the socket is ended and the socket is closed automatically after the currently pending write queue is drained. If `allowHalfOpen` is `true`, the writable side of the socket is kept open, requiring the user to explicitly end it themselves; once the user does end the writable side, the socket is closed automatically after the currently pending write queue is drained.

This is the model I think we should implement, including the optional `allowHalfOpen`.

Reply:
Huh. That seems kind of weird to me. Half-open state is really necessary for most protocols that have a concept of graceful shutdown. For those that don't have any such concept, half-open state typically doesn't occur in the first place since whichever side decides to disconnect will typically close both directions at once.

I guess making `allowHalfOpen` opt-in seems useful to avoid resource leaks in apps that forget to write a close handler? Is that the reasoning?

It's too bad the `ReadableStream` interface doesn't use an `EventTarget` style with a close handler. If it did, then we could say the default behavior of the close handler is to close the opposite direction, and you use `event.preventDefault()` to block that. But given that the style is that `read()` returns a special value to indicate EOS, it's not clear where this "default behavior" would hook in.

I wonder if the resource leak problem is less of an issue for us due to IoContexts -- we'll close all outbound sockets anyway once the IoContext is "done".